### ISSUE TITLE
C++11 support for OpenMS

### ISSUE THREAD

**Issue by hroest** [OpenMS/OpenMS#980]
Currently the main issue with C++11 support for OpenMS is our Python bindings. Specifically, the Python 2.7 binaries for Windows are only built using MSVS 2008 which does not support C++11 [1,2]. RHEL 6.0 is another legacy system that has gcc 4.4 which has partial support for C++11 [5,6].

Python 3.3 and 3.4 seem to use MSVS 2010 [3] which supports a subset of C++11 [4].
Python 3.5 uses MSVS 2015 [3]. Probably Python 3.6 as well, according to [3].

The main obstacle here is Python 2.x on Windows, however people with this configuration will probably be our primary target group for intermediate programmers that want to try out pyOpenMS. We have made progress to support Python 3 #744 and in principle, the technical path to dropping Python 2.x on Windows would be open and replace it with Python 3 support. Other systems such as RHEL also only have partial support for C++11 (gcc 4.4).

Ref:
1. https://mail.python.org/pipermail/python-dev/2012-August/121410.html 
2. http://stackoverflow.com/questions/2676763/what-version-of-visual-studio-is-python-on-my-computer-compiled-with
3. https://docs.python.org/devguide/setup.html#windows
4. http://msdn.microsoft.com/en-us/library/hh567368.aspx
5. https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/6.0_Release_Notes/compiler.html#idp15008256
6. https://gcc.gnu.org/projects/cxx0x.html

**Comment by hroest** [comment]
If we move to C++11 I suggest that the maximal amount of features that we should use is the intersection of MSVS 2010 and gcc 4.4.7 as these are the legacy system that are likely important to continue supporting. gcc because of CentOS systems that at least here in Zurich have a large install base and MSVS2010 because that is the compiler for which Python 3.3 and 3.4 are built with by default on Windows. For the features see this list

- https://gcc.gnu.org/gcc-4.4/cxx0x_status.html
- http://blogs.msdn.com/b/vcblog/archive/2010/04/06/c-0x-core-language-features-in-vc10-the-table.aspx

in short: rvalue yes (partial), auto yes, lambda no, atomic no, threads mostly no, std::move yes (partial)

**Comment by hendrikweisser** [comment]
@hroest:
You mentioned #744 as the main blocker in your first post. You've closed that issue in the meantime - so I suppose Python 3 support in pyOpenMS is in place now?

Interestingly, "Python 3.5 and later use Microsoft Visual Studio 2015" (https://docs.python.org/devguide/setup.html#windows). If we could require that for pyOpenMS, it would give us (almost) full support for C++11 on Windows.

Then we'd have to evaluate whether we can make the move to newer compilers on the UNIX side.

**Comment by cbielow** [comment]
The point for Windows is: MS VS compilers are needed and the non-express Versions cost money. So requiring VS 2012 might offend people who paid for VS 2010 and like their plugins (which VS Express does not support).
Just something to keep in mind.
-- 
Sent via mobile phone.

Am 16. November 2015 11:16:14 MEZ, schrieb hendrikweisser <notifications@github.com>:

**Comment by timosachsenberg** [comment]
I know of none windows developer outside the OpenMS developer community.
Do you know any? We might ask around on the mailing list.

**Comment by hendrikweisser** [comment]
For VS 2015, there's a free version that is quite extensive, as far as I know:
"Visual Studio Community is free for individual developers, open source projects, academic research, education, and small professional teams." (https://www.visualstudio.com/products/visual-studio-community-vs)

**Comment by hroest** [comment]
@hendrikweisser yes, in principle Python 3 support is available, we can compile it for Python 3 on a Linux machine and everything runs. 

also it would mean that we have drop Python 2.x support, at least on Windows. This may or may not be a problem for some people and at least we could say that older versions of pyOpenMS (1.11 and 2.0) will still be available for Windows and Python 2.x.

I think on the Unix side there will be problems with older distributions like RedHat which do not have the new compilers with C++11 support installed. I see quite a few people on the mailing list trying to install OpenMS on these machines and already now it is hard enough but will soon become much harder if there are C++11 features in there which require a new compiler -- see for example this answer here http://unix.stackexchange.com/questions/63587/how-to-install-g-4-7-2-c11-on-centos-5-x  

again, my suggestion to address these issues and some raised by @cbielow is to support the common subset of MSVS2010 and gcc 4.4 which already contains some nice things but of course not all of them. We could enforce this by having builds that use the old compilers.

**Comment by mwalzer** [comment]
[here](https://blogs.msdn.microsoft.com/vcblog/2015/06/19/c111417-features-in-vs-2015-rtm/) is an overview on Microsoft support of C++ versions.
Also see [here](https://msdn.microsoft.com/de-de/library/hh567368.aspx).
Just to have the links...

**Comment by cbielow** [comment]
SeqAn 2.1 (released Feb 2016) **requires** C++11, i.e. VS2015 and GCC 4.9.

However, we currently do not even support VS2015 on Windows (because the contrib, especially Xerces and CoinMP, do not have Solution files available). These libs have also NOT been tested on VS2015 by their developers, i.e. we might run into compiler bugs/wrong includes etc...

So, supporting C++11 and/or updating to SeqAn 2.1 is currently a big problem.

**Comment by hroest** [comment]
Seems like we need to stick with older seqan versions. I don't really like
the idea to require cutting edge stuff from compilers. Many university
computing clusters do not run the newest compilers but go for stability.

H

sent from my smartphone -- you may keep any typos you find.

**Comment by cbielow** [comment]
rough road map from dev meeting:
 - add support (OpenMS and contrib) for VS2015
 - when the BSc Thesis for PeptideIndexer is done, think about adding SeqAn >2.1.x to the contrib.
We could use two seqans in parallel, usage depending on compiler, or switch completely to SeqAn2.1, dropping support for anything lower than GCC4.9, VS2015 (I think thats unlikely)

**Comment by hroest** [comment]
why do we need new Seqan versions? I think dropping support for older gcc and thus dropping support for most cluster / fat nodes out there running redhat is a major issue -- and definitely an issue for me!

**Comment by cbielow** [comment]
@hroest I agree completely.
New SeqAn would be required for a faster PeptideIndexer (how much faster we'll see soon).
The solution to support this would be a fallback to old Seqan if required. ...its extra effort in the build system ofc.

To support a better subset of C11 we could bump Windows to VS 2013. That should not hurt anyone and the gcc 4.4 becomes the reference for c11 support.

**Comment by hroest** [comment]
I like it, we just need to make sure we test gcc 4 builds every night to
ensure compatibility.

sent from my smartphone -- you may keep any typos you find.

**Comment by timosachsenberg** [comment]
@hroest what do you think how long it will take till your cluster can be moved to e.g. gcc 4.6 or even better 4.7? Is this something that likely will never happen or is there a roadmap?
I know building gcc as a user-local build with all its dependencies seems to be a PITA but would this be also an option?

**Comment by hroest** [comment]
It's not my cluster we are talking about but all the people out there (and
on the mailing list). I think I could manage to compile my own gcc,
probably compile my own qt etc. But Already now people have enough trouble
compiling.

So basically we should look at how long individual red Hat systems are
supported if we want to know how long people will use them.

sent from my smartphone -- you may keep any typos you find.

**Comment by timosachsenberg** [comment]
https://access.redhat.com/support/policy/updates/errata gives a nice overview.
Production Phase 2 - which starts 2016 Q2 "New software functionality is not available during this phase." and only security fixes / bug fixes are provided.
Redhat 7 in in Phase 1 which software updates and provides a pretty recent gcc (4.8).

**Comment by hroest** [comment]
https://access.redhat.com/solutions/19458 : 

```
The following gcc packages are available as the default compiler:
    RHEL7 : gcc 4.8.x
    RHEL6 : gcc 4.4.x
    RHEL5 : gcc 4.1.x
    RHEL4 : gcc 3.4.x
    RHEL3 : gcc 3.2.x
```

according to @timosachsenberg link above, RHEL 5 will be around at least until March 31, 2017 so we will have to deal with gcc 4.1 until then and with gcc 4.4 until November 30, 2020

**Comment by hroest** [comment]
so my suggestion would be to go for one more release, OpenMS 2.1 with gcc 4.1 functionality only and then switch the `develop` branch to only support gcc 4.4 and not any older compilers

**Comment by hendrikweisser** [comment]
For Red Hat, recent GCC versions are available in the Developer Toolset, via Software Collections. See [here](http://developerblog.redhat.com/2014/10/30/red-hat-software-collections-1-2-now-ga/) for the announcement and e.g. [here](https://access.redhat.com/documentation/en-US/Red_Hat_Developer_Toolset/3/html-single/User_Guide/index.html#tabl-Red_Hat_Developer_Toolset-About) for more info.

The current GCC version offered for RHEL 6 and 7 is 4.9.2.

**Comment by hroest** [comment]
yes, but does installing them require sudo privileges? will Qt that is already installed on the cluster work or will you have to recompile Qt as well?

**Comment by hendrikweisser** [comment]
> yes, but does installing them require sudo privileges?

Probably. If you don't have those, ask your system administrator to install the package for you - like you already have to for the [other dependencies](http://ftp.mi.fu-berlin.de/pub/OpenMS/release-documentation/html/install_linux.html).

> will Qt that is already installed on the cluster work or will you have to recompile Qt as well?

I haven't tried it, but usually it's not a requirement that libraries have to be built with the same compiler.

**Comment by hroest** [comment]
some update on the Python situation: it seems that > 90% of downloads at PyPI are still Python 2.x (status Feb 2016): 
- http://www.i-programmer.info/news/216-python/9534-python-3-gains-support-and-enthusiasm.html
- https://hynek.me/articles/python3-2016/

on the Linux side, the situation is as follows: most distributions seem to want to move to Python 3 as the default somewhere in 2016-2018
-  https://wiki.openstack.org/wiki/Python3#Status_of_Python_3_in_Linux_distributions
- Ubuntu 16.04 has per default Python 3.5.1 and Python 2 is not installed any more per default: https://wiki.ubuntu.com/XenialXerus/ReleaseNotes#Python_3

**Comment by mortonjt** [comment]
Is there still interested in having Python3 compatibility?  Having Python 3 compatibility could allow for compatibility with [scikit-bio](https://github.com/biocore/scikit-bio) and [qiime2](https://github.com/qiime2/qiime2), would could streamline OpenMS with a whole range of multi-omics analyses.

**Comment by timosachsenberg** [comment]
I think there are good changes that we will support Python3 in the near future. We are moving forward to newer c++ versions and can't won't be able to support old compilers (mainly old windows VS versions). This means that we will most likely only be able to provide Python 3 on these platforms anyways. Right now , we still have some Python 2 code that needs to be ported first so I can't give an exact time frame when to expect this.

**Comment by hroest** [comment]
We currently have full support for Python 3 and also offer packages for 3.3 and 3.4 here on pypi https://pypi.python.org/pypi/pyopenms 

You can also build it yourself, the code is fully Python2 and Python3 compatible. 

Regarding Python 3 adoption, here is a graph on the package level for pypi: http://py3readiness.org/

**Comment by hroest** [comment]
Another, more complete website tracking this is https://caniusepython3.com/  - currently it is at "26670 of 92352 projects compatible." which is only a slight improvement from 9 900 out of 55 000 in April 2014. So it seems while the total number of packages that support Python 3 is increasing, the total number of packages which are Python-2 only is increasing faster in absolute numbers.

**Comment by uweschmitt** [comment]
I don’t expect that Python 2 will disappear in the labs and in many workflows soon. I myself still stick with emzed to Python 2.7 for some reasons and was happy about pyopenms 2.0.1 because of the bug fixes…

Porting existing code from Python2 to Python3 can be pain full, at least the transition to unicode strings as default is not easy in general, I needed some time to get my head around this, and in the web development community this change was difficult. http://lucumr.pocoo.org/2011/12/7/thoughts-on-python3/ is from 2011 but shows what issues come up. The situation gets worse if you consider that many users of pyopenms are not trained computer scientists or might have little programming experience in general.

So dropping support for Python 2.7 might cause some frustration.

I’m aware that getting rid of supporting old MSVC versions would make the build process much easier and new C++ features would be beneficial for the code base, so maybe branching for backporting bug fixes (as Python 2.7 did it) could be a partial solution.

What do you think?

Best
Uwe



--
Dr. Uwe Schmitt
Scientific IT Services
ETH Zürich
Weinbergstr. 11 (WEC D 14)
8092 Zürich
Switzerland
+41 (0) 44 632 06 53

**Comment by timosachsenberg** [comment]
Hi @uweschmitt and @hroest for your input.
The situation is indeed not very satisfying but it is as it is. We will get contributions that require SeqAn > 2 for various reasons and we will try to figure how to deal with the situation. I hope we can keep Python 2 support for quite a while or at least provide backports of bug fixes. That being said, I think we should encourage new users to use Python 3.

**Comment by hroest** [comment]
There are two important points to make

- If we drop support for MSVS2008, it will only mean dropping Python 2 support *on Windows*. All other platforms are fine
- Currently there are several Python 2.x releases of pyopenms which people can use and in general it is not unreasonable for people to expect to move on to a newer version of the software if they are in need of new features (e.g. if they want to use OpenMS 2.x for some version of x than they need to use Python 3 on Windows)

I know this is not perfect but I also see how we cannot be stuck in the past forever. Also I have the feeling that Python 3.x adoption is later among Linux users as many web frameworks need Python 2.x and also many distributions have not switched. Maybe Windows is better? Here Anaconda seems popular and has supported 3.x for long. I also found this http://astrofrog.github.io/blog/2015/05/09/2015-survey-results/ which somewhat underscores this (on Windows its 60:40 for Python2:3 whereas on Linux it was 80:20 but the Windows users were less than 70 people). But if there are 60% Py2 users on Windows, dropping support is still a problem. But I think we should somehow expect this to happen in the next few years.

**Comment by hroest** [comment]
> That being said, I think we should encourage new users to use Python 3.

I agree totally. I think it would be best if we had a transition period where we do Python 2 and Python 3 releases and then see how things go from there.

**Comment by hroest** [comment]
should we close this after #2734
